rafthttp/linkstatus.go

```
@@ -0,0 +1,94 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rafthttp
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+
+	"github.com/coreos/etcd/pkg/types"
+)
+
+var (
+	linkStateMapMu sync.Mutex
+	linkStateMap   = make(map[linkName]linkState)
+)
+
+type linkName struct {
+	local, remote types.ID
+	src           string
+}
+
+type status bool
+
+const (
+	statusOK  status = true
+	statusBad status = false
+)
+
+type linkState struct {
+	status status
+	errm   map[string]bool
+}
+
+func reportLinkStatus(local, remote types.ID, src string, status status, err *opError) {
+	linkStateMapMu.Lock()
+	defer linkStateMapMu.Unlock()
+	name := linkName{local: local, remote: remote, src: src}
+	if _, ok := linkStateMap[name]; !ok {
+		linkStateMap[name] = linkState{status: statusOK, errm: make(map[string]bool)}
+	}
+	if status != linkStateMap[name].status {
+		if strings.HasPrefix(src, "stream") {
+			if status == statusOK {
+				log.Printf("rafthttp: %s from %s to %s becomes active", src, local, remote)
+			} else {
+				log.Printf("rafthttp: %s from %s to %s becomes inactive", src, local, remote)
+			}
+		}
+		linkStateMap[name] = linkState{status: status, errm: make(map[string]bool)}
+	}
+	// TODO: print out the error still after some timeout
```

TODO: do not ignore the error after given timeout